packages/inference/src/lib/makeRequestOptions.ts

```diff
@@ -68,44 +87,52 @@ export async function makeRequestOptions(
 		  ? "credentials-include"
 		  : "none";
 
+	// Make URL
 	const url = endpointUrl
 		? chatCompletion
 			? endpointUrl + `/v1/chat/completions`
 			: endpointUrl
-		: makeUrl({
-				authMethod,
-				chatCompletion: chatCompletion ?? false,
+		: providerConfig.makeUrl({
+				baseUrl:
+					authMethod !== "provider-key"
+						? HF_HUB_INFERENCE_PROXY_TEMPLATE.replace("{{PROVIDER}}", provider)
+						: providerConfig.baseUrl,
 				model,
-				provider: provider ?? "hf-inference",
 				taskHint,
+				chatCompletion,
 		  });
 
-	const headers: Record<string, string> = {};
-	if (accessToken) {
-		if (provider === "fal-ai" && authMethod === "provider-key") {
-			headers["Authorization"] = `Key ${accessToken}`;
-		} else if (provider === "black-forest-labs" && authMethod === "provider-key") {
-			headers["X-Key"] = accessToken;
-		} else {
-			headers["Authorization"] = `Bearer ${accessToken}`;
-		}
-	}
-
-	// e.g. @huggingface/inference/3.1.3
-	const ownUserAgent = `${packageName}/${packageVersion}`;
-	headers["User-Agent"] = [ownUserAgent, typeof navigator !== "undefined" ? navigator.userAgent : undefined]
-		.filter((x) => x !== undefined)
-		.join(" ");
-
+	// Make headers
 	const binary = "data" in args && !!args.data;
+	const headers = providerConfig.makeHeaders({
+		accessToken,
+		authMethod,
+	});
 
+	// Add content-type to headers
 	if (!binary) {
 		headers["Content-Type"] = "application/json";
 	}
 
-	if (provider === "replicate") {
-		headers["Prefer"] = "wait";
-	}
+	// Add user-agent to headers
+	// e.g. @huggingface/inference/3.1.3
+	const ownUserAgent = `${packageName}/${packageVersion}`;
+	const userAgent = [ownUserAgent, typeof navigator !== "undefined" ? navigator.userAgent : undefined]
+		.filter((x) => x !== undefined)
+		.join(" ");
+	headers["User-Agent"] = userAgent;
+
+	// Make body
+	const body = binary
+		? args.data
+		: JSON.stringify(
+				providerConfig.makeBody({
+					args: remainingArgs as Record<string, unknown>,
+					model,
+					taskHint,
+					chatCompletion,
+				})
+		  );
 
```

Comment:
Independent from the changes introduced here: I dislike this binary data special case, it's not obvious from an external pov that passing `data` in arguments will result in the payload being passed as raw bytes in the body.